server/chat-plugins/battlesearch.ts

```diff
@@ -87,159 +157,230 @@ export async function runBattleSearch(userids: ID[], month: string, tierid: ID,
-		buf += `<tr><th>Won</th><th>Lost</th><th>Turns</th></tr>`;
-		for (const battle of outcomes) {
-			const {won, lost, turns} = battle;
-			buf += `<tr><td>${won}</td><td>${lost}</td><td>${turns}</td></tr>`;
-		}
-	}
-	buf += `</tbody></table><br />`;
-	for (const day in data) {
-		const dayStats = data[day];
-		buf += `<p style="text-align:left">`;
-		const {totalWins, totalLosses} = dayStats;
-		buf += `<table style=""><tbody><tr><th colspan="2"><h3 style="margin: 5px auto">${day}</h3>`;
-		buf += `</th></tr><tr><th>Category</th><th>Number</th></tr>`;
-		buf += `<tr><td>Total Battles</td><td>${dayStats.totalBattles}</td></tr>`;
-		for (const id in totalWins) {
-			// hide userids if we're only searching for 1
-			buf += `<tr><td>Total Wins${userids.length > 1 ? ` (${id}) ` : ''}</td><td>${totalWins[id]}</td></tr>`;
-		}
-		for (const id in totalLosses) {
-			buf += `<tr><td>Total Losses${userids.length > 1 ? ` (${id}) ` : ''}</td><td>${totalLosses[id]}</td></tr>`;
-		}
-		if (userids.length < 2) {
-			buf += `<tr><th>Opponent</th><th>Times Battled</th></tr>`;
-			const [userid] = userids;
-			for (const foe in dayStats.timesBattled) {
-				buf += `<tr><td>`;
-				buf += `<a href="/view-battlesearch-${userid}-${foe}--${turnLimit}--${month}--${tierid}" target="replace">${foe}</a>`;
-				buf += `</td><td>${dayStats.timesBattled[foe]}</td></tr>`;
+				const outcomes = results[day].totalOutcomes;
+				if (outcomes) {
+					if (!winnerid) outcomes.push({won: '', lost: '', turns: data.turns, tied: true});
+					else outcomes.push({won: winnerid, lost: loser, turns: data.turns});
+				}
+
+				// we don't want foe data if we're searching for 2 userids
+				const foe = userids.length > 1 ? null : userid === p1id ? p2id : p1id;
+				if (foe) {
+					if (!results[day].timesBattled[foe]) results[day].timesBattled[foe] = 0;
+					results[day].timesBattled[foe]++;
+				}
 			}
 		}
-		buf += `</p><br />`;
+		return results;
+	}
+	async listMonths() {
+		return (await FS('logs/').readdir()).filter(f => f.length === 7 && f.includes('-'));
+	}
+	listTiers(month: string) {
+		return FS(`logs/${month}/`).readdir();
 	}
-	buf += `</tbody></table></div>`;
-	return buf;
 }
 
-async function getBattleSearch(
-	connection: Connection, userids: string[], month: string,
-	tierid: ID, turnLimit?: number
-) {
-	userids = userids.map(toID);
-	const user = connection.user;
-	if (!user.can('forcewin')) return connection.popup(`/battlesearch - Access Denied`);
+function monthSort(aKey: string, bKey: string) {
+	const a = aKey.split('-').map(n => parseInt(n));
+	const b = bKey.split('-').map(n => parseInt(n));
+	if (a[0] !== b[0]) return b[0] - a[0];
+	return b[1] - a[1];
+}
 
-	const response = await PM.query({userids, turnLimit, month, tierid});
-	connection.send(buildResults(response, userids as ID[], month, tierid, turnLimit));
+function toDate(str: string) {
+	return Chat.toTimestamp(new Date(Number(str)));
 }
 
```

Comment:
Have you tested these queries on a large database of battles? Are they performant? It would be nice to get battlesearch right this time; right now, staff rely on an external tool to manually check battle logs by format and date.